ftlib/commlib/nccl/impl.py

```diff
@@ -36,75 +35,92 @@ class NCCL(BasicCommLib):
     def __init__(
         self,
         grad_sync_timeout=10,
-        shared_path="/crystal",
+        shared_path=None,
         filename="nccl_id_file",
         max_try=30,
         wait_time=5,
     ):
         self.type = "NCCL"
         self.grad_sync_timeout = grad_sync_timeout
+        if shared_path is None:
+            self.shared_path = os.getenv("NCCL_ID_DIR", "/crystal")
         self.shared_path = shared_path
         self._nccl_context = fault_tolerant_lib.nccl_context()
         self._nccl_id_filename = filename
         self._max_try = max_try
         self._wait_time = wait_time
+        self._default_timeout = wait_time
 
     @BasicCommLib.register_api
     def grad_sync_done(self):
-        try:
-            self.allreduce(np.array(range(10)).astype(np.float))
-        except Exception as e:
-            logging.warning(str(e))
-            return CommLibStatus.FAIL
-        return CommLibStatus.SUCCESS
+        raise NotImplementedError
 
     @BasicCommLib.register_api
-    def broadcast(self, data, root):
-        logging.debug("broadcasting: " + str(data))
+    def broadcast(self, data, root_rank, timeout=None):
```

Comment:
The args are different than other commlib implementations (due to async nature) but how do we plan to use this with ftlib?

Reply:
It will be an issue if user switch from *NCCL implementation* to *PyTorch implementation*. But it shouldn't cause issue when converting to *NCCL* from *PyTorch*.

I think we can add `*argc, **kwargs` to the args in PyTorch.